src/api_server/src/request/vsock.rs

```diff
@@ -4,32 +4,71 @@
 use super::super::VmmAction;
 use crate::parsed_request::{Error, ParsedRequest};
 use crate::request::Body;
+use logger::{IncMetric, METRICS};
 use vmm::vmm_config::vsock::VsockDeviceConfig;
 
 pub(crate) fn parse_put_vsock(body: &Body) -> Result<ParsedRequest, Error> {
-    Ok(ParsedRequest::new_sync(VmmAction::SetVsockDevice(
-        serde_json::from_slice::<VsockDeviceConfig>(body.raw()).map_err(Error::SerdeJson)?,
-    )))
+    METRICS.put_api_requests.vsock_count.inc();
+    let vsock_cfg = serde_json::from_slice::<VsockDeviceConfig>(body.raw()).map_err(|e| {
+        METRICS.put_api_requests.vsock_fails.inc();
+        Error::SerdeJson(e)
+    })?;
+
+    // Check for the presence of deprecated `vsock_id` field.
+    let mut deprecation_message = None;
+    if vsock_cfg.vsock_id.is_some() {
+        // vsock_id field in request is deprecated.
+        METRICS.deprecated_api.deprecated_http_api_calls.inc();
+        deprecation_message = Some("PUT /vsock: vsock_id field is deprecated.");
+    }
+
+    // Construct the `ParsedRequest` object.
+    let mut parsed_req = ParsedRequest::new_sync(VmmAction::SetVsockDevice(vsock_cfg));
+    // If `vsock_id` was present, set the deprecation message in `parsing_info`.
+    if let Some(msg) = deprecation_message {
+        parsed_req.parsing_info().append_deprecation_message(msg);
+    }
+
+    Ok(parsed_req)
```

Comment:
1. Initially that's what I wanted to do, but you can't since `vsock_cfg` is moved by the time you call `vsock_cfg.vsock_id.is_some()`.
2. I think it's going to be a very rare occurrence to have multiple fields deprecated in a single endpoint. I don't think it's worth complicating the interface for this case. Contributors should decide the format of the message appended.
3. It's not a 1:1 mapping. You can have 2 field deprecated in a single endpoint, but the metric should be incremented only once. For this reason, I would not add the metric `inc` inside `append_deprecation_message`. Maybe it's worth adding a code comment about this, WDYT?

Reply:
1. Sorry, didn't realize.
2. Ok, works for me.
3. Yes, a comment would be really useful.